src/freenet/node/DNSRequester.java

```diff
@@ -54,25 +62,51 @@ public void run() {
     }
 
     private void realRun() {
-        PeerNode[] nodes = node.getPeers().myPeers();
-        long now = System.currentTimeMillis();
-        if((now - lastLogTime) > 1000) {
-        	if(logMINOR)
-        		Logger.minor(this, "Processing DNS Requests (log rate-limited)");
+        // run DNS requests for not recently checked, unconnected
+        // nodes to avoid the coupon collector's problem
+        PeerNode[] nodesToCheck = Arrays.stream(node.getPeers().myPeers())
+            .filter(peerNode -> !peerNode.isConnected())
+            // identify recent nodes by location, because the exact location cannot be used twice
+            // (that already triggers the simplest pitch black attack defenses)
+            // Double may not be comparable in general (floating point),
+            // but just checking for equality with itself is safe
+            .filter(peerNode -> !recentNodeIdentitySet.contains(peerNode.getLocation()))
+            .toArray(PeerNode[]::new);
+
+        if(logMINOR) {
+            long now = System.currentTimeMillis();
+            if((now - lastLogTime) > 100) {
+            		Logger.minor(this, "Processing DNS Requests (log rate-limited)");
+            }
             lastLogTime = now;
         }
-        for(PeerNode pn: nodes) {
-            //Logger.minor(this, "Node: "+pn);
-            if(!pn.isConnected()) {
-                // Not connected
-                // Try new DNS lookup
-            	//Logger.minor(this, "Doing lookup on "+pn+" of "+nodes.length);
-                pn.maybeUpdateHandshakeIPs(false);
+        // check a randomly chosen node that has not been checked
+        // recently to avoid sending bursts of DNS requests
+        int unconnectedNodesLength = nodesToCheck.length;
+        PeerNode pn = nodesToCheck[node.getFastWeakRandom().nextInt(unconnectedNodesLength)];
+        if (unconnectedNodesLength < 5) {
+            // no need for optimizations: just clear all state
+            recentNodeIdentitySet.clear();
+            recentNodeIdentityQueue.clear();
+        } else {
+            // do not request this node again,
+            // until at least 80% of the other unconnected nodes have been checked
+            recentNodeIdentitySet.add(pn.getLocation());
+            recentNodeIdentityQueue.offerFirst(pn.getLocation());
+            while (recentNodeIdentityQueue.size() > (0.81 * unconnectedNodesLength)) {
+                recentNodeIdentitySet.remove(recentNodeIdentityQueue.removeLast());
             }
         }
+        //Logger.minor(this, "Node: "+pn);
+
+        // Try new DNS lookup
+        //Logger.minor(this, "Doing lookup on "+pn+" of "+nodesToCheck.length);
```

Comment:
Can we get rid of the commented out debug code? :smile:

Reply:
Deleted ☺